substrate/runtime/primitives/src/int128.rs

```diff
@@ -0,0 +1,148 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Substrate Demo.
+
+// Substrate Demo is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate Demo is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate Demo.  If not, see <http://www.gnu.org/licenses/>.
+
+//! 128-bit shims so that we can make it `serde::Serialize`-able, since Serde is borked for
+//! the new 128-bit datatypes.
+
+#[cfg(feature = "std")]
+use serde::ser::{Serialize, Serializer, SerializeTuple};
+use codec::{Slicable, Input};
+use integer_sqrt::IntegerSquareRoot;
+use num_traits::{Zero, One, Bounded};
+use rstd::ops::{Add, Sub, Mul, Div, Rem, AddAssign, SubAssign, MulAssign, DivAssign, RemAssign};
+use traits::As;
+
+/// A 128-bit uint shim.
+#[derive(Debug, Eq, PartialEq, PartialOrd, Ord, Clone, Copy, Default)]
+pub struct U128(u128);
+
+/// A 128-bit int shim.
+#[derive(Debug, Eq, PartialEq, PartialOrd, Ord, Clone, Copy, Default)]
+pub struct I128(i128);
+
+macro_rules! impl_as {
+	( $f:ident, $i:ty ) => {
+		impl_as!($f, $i: u8, u16, u32, u64, u128, usize, i8, i16, i32, i64, i128, isize);
+	};
+	( $f:ident, $i:ty : $t:ty $(, $rest:ty)* ) => {
+		impl As<$t> for $f {
```

Comment:
why not just `From` and `Into`?

Reply:
the semantics are different for some types that we don't have control over.

Reply:
Still, this `As` trait is just `From + Into`, isn't it?

Reply:
Well, it is used a bit like `From` and `Into`. Except that it can convert between numeric types like the `as` keyword.

Reply:
so it's just an escape hatch around the orphan rules for primitives? it could be that `TryFrom/TryInto` will be implemented for all pairs in the future.

Reply:
yup